File: WeatheriOS/Weather/Services/WeatherService.swift
//
//  WeatherService.swift
//  Weather
//
//  Created by Rick Pasetto on 12/21/25.
//

import Foundation

class WeatherService {
    private let baseURL = "https://api.weather.gov"
    
    func fetchWeather(for city: City) async throws -> WeatherInfo {
        let points = try await fetchGridPoints(latitude: city.latitude, longitude: city.longitude)
        let forecast = try await fetchForecast(url: points.properties.forecast)
        let currentWeather = try? await fetchCurrentWeather(
            gridId: points.properties.gridId,
            gridX: points.properties.gridX,
            gridY: points.properties.gridY
        )
        
        return WeatherInfo(
            city: city,
            currentTemperature: currentWeather?.temperature,
            currentCondition: currentWeather?.condition,
            forecast: forecast.properties.forecastPeriods
        )
    }
    
    private func makeRequest(url: URL) async throws -> (Data, URLResponse) {
        var request = URLRequest(url: url)
        request.setValue("WeatherApp (iOS)", forHTTPHeaderField: "User-Agent")
        return try await URLSession.shared.data(for: request)
    }
    
    private func fetchGridPoints(latitude: Double, longitude: Double) async throws -> PointsResponse {
        let url = URL(string: "\(baseURL)/points/\(latitude),\(longitude)")!
        let (data, _) = try await makeRequest(url: url)
        return try JSONDecoder().decode(PointsResponse.self, from: data)
    }
    
    private func fetchForecast(url: URL) async throws -> ForecastResponse {
        let (data, _) = try await makeRequest(url: url)
        return try JSONDecoder().decode(ForecastResponse.self, from: data)
    }
    
    private func fetchCurrentWeather(gridId: String, gridX: Int, gridY: Int) async throws -> (temperature: Int?, condition: String?)? {
        let url = URL(string: "\(baseURL)/gridpoints/\(gridId)/\(gridX),\(gridY)/stations")!
        let (data, _) = try await makeRequest(url: url)
        let stationsResponse = try JSONDecoder().decode(StationsResponse.self, from: data)
        
        guard let stationId = stationsResponse.features.first?.properties.stationIdentifier else {
            return nil
        }
        
        let observationURL = URL(string: "\(baseURL)/stations/\(stationId)/observations/latest")!
        let (obsData, _) = try await makeRequest(url: observationURL)
        let observation = try JSONDecoder().decode(ObservationResponse.self, from: obsData)
        
        let temperature = observation.properties?.temperature?.value.map { Int($0) }
        let condition = observation.properties?.textDescription
        
        return (temperature: temperature, condition: condition)
    }
}

struct StationsResponse: Codable {
    let features: [StationFeature]
}

struct StationFeature: Codable {
    let properties: StationProperties
}

struct StationProperties: Codable {
    let stationIdentifier: String
}

File: WeatherRN/src/services/weatherService.js
const BASE_URL = 'https://api.weather.gov';

const headers = {
  'User-Agent': 'WeatherApp (ReactNative)',
};

async function fetchGridPoints(latitude, longitude) {
  const response = await fetch(
    `${BASE_URL}/points/${latitude},${longitude}`,
    { headers }
  );
  if (!response.ok) throw new Error('Failed to fetch grid points');
  return response.json();
}

async function fetchForecast(forecastUrl) {
  const response = await fetch(forecastUrl, { headers });
  if (!response.ok) throw new Error('Failed to fetch forecast');
  return response.json();
}

async function fetchCurrentObservation(gridId, gridX, gridY) {
  try {
    const stationsUrl = `${BASE_URL}/gridpoints/${gridId}/${gridX},${gridY}/stations`;
    console.log(`[DEBUG] Fetching stations: ${stationsUrl}`);
    
    const stationsResponse = await fetch(stationsUrl, { headers });
    if (!stationsResponse.ok) {
      console.warn(`[DEBUG] Stations request failed: ${stationsResponse.status}`);
      return { error: `Stations API error: ${stationsResponse.status}` };
    }
    
    const stations = await stationsResponse.json();
    const stationId = stations.features?.[0]?.properties?.stationIdentifier;
    if (!stationId) {
      console.warn(`[DEBUG] No station found for ${gridId}/${gridX},${gridY}`);
      return { error: 'No weather station found' };
    }
    console.log(`[DEBUG] Found station: ${stationId}`);

    const obsUrl = `${BASE_URL}/stations/${stationId}/observations/latest`;
    const obsResponse = await fetch(obsUrl, { headers });
    if (!obsResponse.ok) {
      console.warn(`[DEBUG] Observation request failed: ${obsResponse.status}`);
      return { error: `Observation API error: ${obsResponse.status}` };
    }

    const observation = await obsResponse.json();
    const tempCelsius = observation.properties?.temperature?.value;
    const condition = observation.properties?.textDescription || null;
    
    console.log(`[DEBUG] Station ${stationId}: temp=${tempCelsius}, condition=${condition}`);
    
    // If station returns null data, it's not reporting
    if (tempCelsius == null && !condition) {
      return { error: `Station ${stationId} not reporting` };
    }
    
    return {
      temperature: tempCelsius != null ? Math.round(tempCelsius * 9/5 + 32) : null,
      condition,
    };
  } catch (e) {
    console.error(`[DEBUG] fetchCurrentObservation error:`, e.message);
    return { error: e.message };
  }
}

export async function fetchWeather(city) {
  const points = await fetchGridPoints(city.latitude, city.longitude);
  const { gridId, gridX, gridY, forecast: forecastUrl } = points.properties;
  
  const [forecastData, current] = await Promise.all([
    fetchForecast(forecastUrl),
    fetchCurrentObservation(gridId, gridX, gridY),
  ]);

  return {
    id: city.id,
    city,
    currentTemperature: current?.temperature ?? null,
    currentCondition: current?.condition ?? null,
    currentError: current?.error ?? null,
    forecast: forecastData.properties.periods.slice(0, 5).map((p, i) => ({
      id: `${city.id}-${i}`,
      name: p.name,
      temperature: p.temperature,
      temperatureUnit: p.temperatureUnit,
      shortForecast: p.shortForecast,
    })),
  };
}


